lib/endoscope-abi/src/lib.rs

```diff
@@ -0,0 +1,65 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! endoscope-abi
+//!
+//! This crate documents the interface of the program that the RoT injects into
+//! the SP that measures the SP flash contents.
+//!
+//! When SP reset is detected, the RoT halts the SP, injects the endoscope program
+//! into the SP RAM and runs it. The endoscope program will measure the entire
+//! active flash bank and deposit the resulting Sha3-256 hash into the Shared
+//! structure. The RoT polls the STM32 debug module waiting for a halt or timeout.
+//! The valid measurement is retrieved and recorded if availalble.
+//!
+
+#![no_std]
+
+use zerocopy::*;
+
+#[repr(u32)]
+pub enum State {
+    #[allow(dead_code)]
+    Preboot = 0,
+    Running = 0x1de6060,
+    Done = 0x1dec1a0,
+}
+
+#[derive(FromBytes, AsBytes, Copy, Clone)]
+#[repr(C, packed)]
+pub struct Shared {
+    pub magic: u32,
+    pub state: u32,
+    pub start: u32,
+    pub len: u32,
+    pub digest: [u8; 256 / 8],
+}
+
+impl Shared {
+    pub const MAGIC: u32 = 0x1de2019;
+    pub const STATE_PREBOOT: u32 = 0;
+    /// The program main routine has started.
+    pub const STATE_RUNNING: u32 = 0x1de6060;
+    /// The program main routine finished. Shared::digest is valid.
+    pub const STATE_DONE: u32 = 0x1dec1a0;
+
+    pub fn parse(bytes: &[u8]) -> Option<&Self> {
+        if let Some((layout, _)) =
+            LayoutVerified::<&[u8], Self>::new_from_prefix(bytes)
+        {
+            let shared: &Shared = layout.into_ref();
+            Some(shared)
+        } else {
+            None
+        }
+    }
+}
+
+// Symbols relied on in the endoscope.elf file.
```

Comment:
It feels a little weird to have these symbol strings in the `endoscope-abi` crate, since they're not actually part of the embedded ABI.  Should they be constants in the `build.rs` instead?

Reply:
Thanks!

This is some speculative generalization that can be eliminated.

My thinking is that part of the ABI is where to inject the endoscope blob into the SP's memory, what its entry address is and where to find the results.
That allows good flexibility for developing the blob independent from the RoT code. So, the "private" symbol names need to be bound somehow to those used by the `swd` task.

Since drv/lpc55-swd/build.rs is the only user and endoscope is the only injected blob used in this way, I'm happy to move these to build.rs.